gzip_norace_test.go

```diff
@@ -0,0 +1,81 @@
+// These tests are skipped when the race detector (-race) is on
+// +build !race
+
+package pgzip
+
+import (
+	"bytes"
+	"io/ioutil"
+	"runtime"
+	"runtime/debug"
+	"testing"
+)
+
+// Test that the sync.Pools are working properly and we are not leaking buffers
+// Disabled with -race, because the race detector allocates a lot of memory
+func TestAllocations(t *testing.T) {
+
+	w := NewWriter(ioutil.Discard)
+	w.SetConcurrency(100000, 10)
+	data := bytes.Repeat([]byte("TEST"), 41234) // varying block splits
+
+	// Prime the pool to do initial allocs
+	for i := 0; i < 10; i++ {
+		_, _ = w.Write(data)
+	}
+	_ = w.Flush()
+
+	allocBytes := AllocBytesPerRun(1000, func() {
+		_, _ = w.Write(data)
+	})
+	t.Logf("Allocated %.0f bytes per Write on average", allocBytes)
+
+	// Locally it still allocates 660 bytes, which can probably be further reduced,
+	// but it's better than the 175846 bytes before the pool release fix this tests.
+	// TODO: Further reduce allocations
+	if allocBytes > 10240 {
+		t.Errorf("Write allocated too much memory per run (%.0f bytes), Pool used incorrectly?", allocBytes)
```

Comment:
For the failures it reports `gzip_test.go:637: Write allocated too much memory per run (18446744073709436 bytes), Pool used incorrectly?`

That is 18446.7TB - I think you would be hard pressed to allocated that much, so I suspect there is something wrong.

Reply:
Are you sure this happened with the latest version? I noticed this in earlier test runs and 799eba1c20806ffaa5c2a8ec998ba1d062e9f062 fixes this. Before that commit I was using `.Alloc` instead of `.TotalAlloc`.

From the `MemStats` comments:

> TotalAlloc increases as heap objects are allocated, but unlike Alloc and HeapAlloc, it does not decrease when objects are freed.

Reply:
Yeah, seems I mixed up the runs. Anyway, merging master should fix the writer.